### DISCUSSION TITLE
Mock function supported

### DISCUSSION THREAD

**Discussion by gpacurar53** [pactumjs/pactum#93] [Q&A] [answered]
Hello,

Does pactumjs supports mock functions?

**Comment by leelaprasadv** [comment]
Hey! 
Can you please provide more details about what mock functions are you looking for? Maybe with an example?

Pactum library supports mock http server to mock out any external dependencies while testing an application.
More details here: 
https://pactumjs.github.io/#/mock-server

**Comment by gpacurar53** [comment]
I am trying to mock a function not a server, and store it in a variable, and do some assertion in the end, I have found an example with jest as you can see below.

```
const mockCallback = jest.fn(x => 42 + x);
forEach([0, 1], mockCallback);

// Do some assertion 
expect(mockCallback.mock.calls.length).toBe(2);
```

**Comment by ASaiAnudeep** [comment, marked as answer]
###### Welcome to PactumJS Discussions & thanks a lot for using this library.

Pactum doesn't support mocking functions. Usually mocking independent functions happens at unit test level. If you are writing unit tests, we can use libraries like [sinon](https://www.npmjs.com/package/sinon) or jest for mocking.

If you are writing API tests, we generally mock the external dependencies. I do understand that at some point particular dependencies cannot be mocked. As a work around, I prefer using env variables to control the behaviour. For example

```js

function getSecret() {
  if (process.env.NODE_ENV === 'test') { return 'secret'; }

  // or else actual code will be executed to fetch the secret

}
```

Running the server in test environment.

```sh
export NODE_ENV=test && node server.js
``` 

The challenge with the above code is, it is not dynamic.

**Comment by martynaol** [comment]
Hello ASaiAnudeep,
I encounter an issue that I mocked request and response but during the test, pactum take real data not these mocked ones.  I have file spec.integration_handlers where I have added interaction, in the test spec in the before a section I use
 ```js  
 mock.addInteraction('getLayouts');
 mock.start();
 ```

And when I have step where I want to get layout
```js
Given('I have the layout', async () => {
  layoutId = await pactum.spec().get('/layouts').expectStatus(200).returns('data[0].id');
  console.log('layout: ', layoutId);
});
```
layoutId is not taken from the mocked response but I had rlayout:  1ca5b9a0-b009-11ec-a376-5de1ad1d5b31

```js
handler: 'handler.addInteractionHandler('getLayouts', () => {
  return {
    strict: false,
    request: {
      method: 'GET',
      path: `/layouts`,
    },
    response: {
      status: 200,
      body: {
        data: [
          {
           
            id: '5de65005-8c14-4459-9a5b-44b70b6f8047',
            industryId: '2e9d94ad-db1d-422d-b7e1-f56afad49493',
            name: 'Layout test',
            usecaseId: 'd855d913-4402-4e2a-bba7-38cf76b50427',
          },
        ],
      },
    },
  };`
```
I have information that `Mock server is listening on http://0.0.0.0:9393`. I changed port, added the host, but  without success

**Reply by ASaiAnudeep** [reply]
Include the entire URL in your spec.

```js
Given('I have the layout', async () => {
  layoutId = await pactum.spec().get('http://localhost:9393/layouts').expectStatus(200).returns('data[0].id');
  console.log('layout: ', layoutId);
});
```

**Reply by martynaol** [reply]
I don't get it, sorry.  So inside the test, do I need to define URL to mock the server? I already have set   `pactum.request.setBaseUrl(.BASE_URL);`, and what if I run the tests on various environments and I only mock request&response when I run in locally ?

**Reply by ASaiAnudeep** [reply]
Hey @martynaol,

Could you please elaborate your use case by creating a new discussion by including more example code.
